<h1>Credit Card Fraud Detction Using Adavanced Classifcation Algorithims</h1></br>
<p>In this project I would be comparing 2 advanced classifaction algoritims and checking out which of the performs best</p>
</br>
<h3>The two algorithims Includes</h3>
<ol>
  <li>Random Forest</li>
  <li>Gradient Boosting Machines (GBM)</li>
 </ol>


In [1]:
# Firstly I would have to import all of the necessary pakeages 
import numpy as np
import pandas as pd 

In [2]:
# Load the dataset
def load_dataset():
    # This Load the dataset here with your actual file path
    file_path = r"C:\Users\Awomodu David\Downloads\Creditcard_data\creditcard.csv"
    data = pd.read_csv(file_path)
    return data

In [3]:
# Let us see what the fist 5 row of the file looks like using the head function 
data_set = load_dataset()
data_set.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

<h3>Link to the data set on kagle</h3>
<p>You can get the link to the dataset <a href="https://www.kaggle.com/mlg-ulb/creditcardfraud/download">here</a>
</p>

In [4]:
# Let us get the shap and the descrption of the data using the shape and describe methods
data_set = load_dataset()
print(data_set.shape)
print(data_set.describe)

(284807, 31)
<bound method NDFrame.describe of             Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462

In [5]:
# This part of the code is to determine number of fraud cases in the dataset
data = load_dataset()
fraud_case = data[data['Class']==1]
valid_transaction = data[data['Class']==0]
outlierFraction = len(fraud_case)/float(len(valid_transaction))
print(outlierFraction)
print('Fraud cases: {}'.format(len(data[data['Class']==1])))
print('Valid Transaction: {}'.format(len(data[data['Class']==0])))

0.0017304750013189597
Fraud cases: 492
Valid Transaction: 284315


<h3>Result obtained after determinng the number of fraud cases and legit transactions</h3>
<ol>
  <li>Fraud case  is 492</li>
  <li>Legitimate transactions is 284315</li>
</ol>


In [6]:
# for this part, we use the anount and describe method to print the detail amount of the fraudulent transaction
print("Amount details of the fraudulent transaction")
fraud_case.Amount.describe()

Amount details of the fraudulent transaction


count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [7]:
# This line of code divide the X and Y from the dataset 
data = load_dataset()
X = data.drop(['Class'], axis = 1)
Y = data['Class']
print(X.shape)
print(Y.shape)
# getting just teh value for the sake of processing 
# it is a numpy array with no colums 
x_data = X.values
y_data = Y.values 

(284807, 30)
(284807,)


In [10]:
# Using Sk learn to split data into training and testing sets 
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score


<h2>Building and comparing the models</h2></br>


In [11]:
# Prepare data for modeling
def prepare_data(data_set):
    # Split data into features (X) and target variable (y)
    X = dataset.drop('Class', axis=1) 
    y = dataset['Class']
    
    # Split the data into train and test sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    return X_train, X_test, y_train, y_test


In [13]:
# Build and evaluate the Random Forest model
def random_forest(X_train, X_test, y_train, y_test):
    rf_model = RandomForestClassifier(random_state=42)
    rf_model.fit(X_train, y_train)
    y_pred = rf_model.predict(X_test)
    
    # Evaluate model performance
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    
    return accuracy, precision, recall, f1

In [14]:
# Build and evaluate the Gradient Boosting Machines (GBM) model
def gbm(X_train, X_test, y_train, y_test):
    gbm_model = GradientBoostingClassifier(random_state=42)
    gbm_model.fit(X_train, y_train)
    y_pred = gbm_model.predict(X_test)
    
    # Evaluate model performance
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    
    return accuracy, precision, recall, f1

In [16]:
dataset = load_dataset()
    
    
X_train, X_test, y_train, y_test = prepare_data(dataset)
    
    # Build and evaluate models
rf_results = random_forest(X_train, X_test, y_train, y_test)
gbm_results = gbm(X_train, X_test, y_train, y_test)

    
    # Display results
print("Random Forest Results:")
print("Accuracy:", rf_results[0])
print("Precision:", rf_results[1])
print("Recall:", rf_results[2])
print("F1 Score:", rf_results[3])
print()
    
print("Gradient Boosting Machines Results:")
print("Accuracy:", gbm_results[0])
print("Precision:", gbm_results[1])
print("Recall:", gbm_results[2])
print("F1 Score:", gbm_results[3])
print()
    


Random Forest Results:
Accuracy: 0.9995611109160493
Precision: 0.974025974025974
Recall: 0.7653061224489796
F1 Score: 0.8571428571428571

Gradient Boosting Machines Results:
Accuracy: 0.9989466661985184
Precision: 0.7375
Recall: 0.6020408163265306
F1 Score: 0.6629213483146067



<style>
        table {
            border-collapse: collapse;
            width: 80%; /* Adjust the width of the table */
            margin: 20px auto; /* Center the table */
        }

        th, td {
            border: 1px solid #dddddd;
            text-align: left;
            padding: 12px; /* Increase padding for better readability */
        }

        th {
            background-color: #f2f2f2;
        }
</style>

    
<h2>Model Comparison</h2>

<table>
    <tr>
        <th>Model</th>
        <th>Accuracy</th>
        <th>Precision</th>
        <th>Recall</th>
        <th>F1 Score</th>
    </tr>
    <tr>
        <td>Random Forest</td>
        <td>0.9995611109160493</td>
        <td>0.974025974025974</td>
        <td>0.7653061224489796</td>
        <td>0.8571428571428571</td>
    </tr>
    <tr>
        <td>Gradient Boosting Machines</td>
        <td>0.9989466661985184</td>
        <td>0.7375</td>
        <td>0.6020408163265306</td>
        <td>0.6629213483146067</td>
    </tr>
</table>

<p><strong>Conclusion:</strong> Based on the comparison of accuracy, precision, recall, and F1 score, the Random Forest model performs better overall compared to the Gradient Boosting Machines model.</p>


